lib/pbench/server/api/resources/query_apis/datasets_owner.py

```diff
@@ -0,0 +1,101 @@
+from logging import Logger
+from typing import Iterator
+
+from pbench.server import JSONOBJECT, OperationCode, PbenchServerConfig
+from pbench.server.api.resources import (
+    ApiAuthorizationType,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+    Parameter,
+    ParamType,
+    Schema,
+)
+from pbench.server.api.resources.query_apis import ApiContext, ElasticBulkBase
+from pbench.server.database.models.datasets import Dataset
+
+
+class DatasetOwner(ElasticBulkBase):
+    """
+    Change the "owner" authorization of a Pbench dataset which is either public or
+    needs change in ownership, by modifying the "authorization": {"owner": owner_id}
+    subdocument of each Elasticsearch document associated with the specified dataset.
+
+    POST /api/v1/datasets/change_owner/{dataset}?owner=user
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            ApiSchema(
+                ApiMethod.POST,
+                OperationCode.UPDATE,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True)
+                ),
+                query_schema=Schema(Parameter("owner", ParamType.USER, required=True)),
+                authorization=ApiAuthorizationType.ADMIN,
+            ),
+            action="update",
+            require_stable=True,
+            require_map=True,
+        )
+
+    def generate_actions(
+        self,
+        params: ApiParams,
+        dataset: Dataset,
+        context: ApiContext,
+        map: dict[str, list[str]],
+    ) -> Iterator[dict]:
+        """
+        Generate a series of Elasticsearch bulk update actions driven by the
+        dataset document map.
+
+        Args:
+            params: API parameters
+            dataset: the Dataset object
+            context: CONTEXT to pass to complete
+            map: Elasticsearch index document map
+
+        Returns:
+            A generator for Elasticsearch bulk update actions
+        """
+
+        owner_id = params.query["owner"]
+        context["owner_id"] = owner_id
+
+        self.logger.info(
+            "Changing owmership operation for dataset {} to user {}", dataset, owner_id
+        )
+
+        for index, ids in map.items():
+            for id in ids:
+                yield {
+                    "_op_type": self.action,
+                    "_index": index,
+                    "_id": id,
+                    "doc": {"authorization": {"owner": owner_id}},
+                }
```

Comment:
Yeah, a lot of the bulk operation subclasses could be consolidated. I've thought about this several times and put it off because what I've wanted to do (and always decided we weren't really ready, but maybe we are) was to use this set as an entry to the V2 resource API with `POST ?access=&owner=` and `DELETE` instead of having three separate URIs for publish, delete, and "change owner". I'm increasing starting to think it may be worth at least merging publish and "change owner" just so we don't have to *name* the new one!

And if we threw in `DELETE` (which should be straightforward) these three would collapse to one class. I kinda like that; although whether we should make Siddardh go that far in the PR is a different question. I'd be happy to take that as a followup.